### PR TITLE
Test SQLFile with all the possible (handled) use cases

### PR DESCRIPTION
**Pull request by frapontillo**
#### Problem

`SQLFile` handles a good part of the library behaviour, collecting all the statements needed to build the in-memory database. That database is later needed to generate the access Java classes.

If SQL files are incorrectly parsed, the generated classes won't match.

#### Solution

This PR adds the `SQLFileTest` spec, defining all of the handled use cases:

- middle case, two statements separated by a new line char
- edge case, two statements surrounded by spaces
- edge case, two statements separated by line comments
- edge case, two statements separated by block comments
- edge case, a statement with line comments in it
- edge case, a statement with block comments in it

##### Test(s) added 

Duh

##### Screenshots

NOPE

###### Paired with 

/foreveralone

### CODE REVIEW

**Comment by frapontillo** [comment]
Side note: this is just the first PR (NCU module 3 compatible, LOL) for a code improvement that will add more tests. These tests would break with the current implementation of `SQLFile` and will therefore be added with an improved `SQLFile`.

File: analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import java.io.StringReader;
+import java.util.List;
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+
+public class SQLFileTest {
+    private final static String STATEMENT_CREATE_TABLE = "CREATE TABLE banana(_id INTEGER);";
```
Comment (rock3r) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: image

File: analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    private final static String STATEMENT_ALTER_TABLE = "ALTER TABLE banana ADD COLUMN colour TEXT;";
+    private final static String NEW_LINE = "\n";
+    private final static String SPACE = " ";
```
Comment (takecare) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: i'm always a bit torn with this... i used to do it but now not so sure if it doesn't make it harder to read

Comment (frapontillo) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: It is a constant in the `SQLFile` class as well, as it is used to concatenate multiple lines of a statement, so I figured I'd declare it here for the same reason.

Comment (rock3r) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: I'd take it out in both places `</my2cents>`

File: analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+        String[] actual = getStatementsFromFile(file);
+
+        assertArrayEquals(actual, EXPECTED_STATEMENTS);
+    }
+
+    @Test
+    public void givenSQLFileWithLineComments_whenParseAndGetStatements_thenIgnoreComments() throws IOException {
+        SQLFile file = givenSQLFileParsedFromString(
+                STATEMENT_CREATE_TABLE + SPACE + LINE_COMMENT + NEW_LINE + STATEMENT_ALTER_TABLE + SPACE + LINE_COMMENT);
```
Comment (rock3r) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: The closed parenthesis (here and below) should go on a newline

Comment (frapontillo) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: f7dad9e1e0332f27d8006cb087416d6dd44a9925

File: analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    private final static String STATEMENT_FOR_COMMENTS_WITHIN_TESTS_1 = "CREATE TABLE";
+    private final static String STATEMENT_FOR_COMMENTS_WITHIN_TESTS_2 = "banana(_id INTEGER);";
+
+    @Test
+    public void givenSQLFile_whenParseAndGetStatements_thenGetCorrectStatements() throws IOException {
+        SQLFile file = givenSQLFileParsedFromString(STATEMENT_CREATE_TABLE + NEW_LINE + STATEMENT_ALTER_TABLE);
+
+        String[] actual = getStatementsFromFile(file);
+
+        assertArrayEquals(actual, EXPECTED_STATEMENTS);
```
Comment (rock3r) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: No fest assertions?

File: analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    private final static String SPACE = " ";
+    private final static String[] EXPECTED_STATEMENTS = new String[]{STATEMENT_CREATE_TABLE, STATEMENT_ALTER_TABLE};
+    private final static String LINE_COMMENT = "-- this is a line comment";
+    private static final String BLOCK_COMMENT = "/* this is a \n block comment \n spanning over \n multiple lines */";
+    private final static String STATEMENT_FOR_COMMENTS_WITHIN_TESTS_1 = "CREATE TABLE";
+    private final static String STATEMENT_FOR_COMMENTS_WITHIN_TESTS_2 = "banana(_id INTEGER);";
+
+    @Test
+    public void givenSQLFile_whenParseAndGetStatements_thenGetCorrectStatements() throws IOException {
+        SQLFile file = givenSQLFileParsedFromString(STATEMENT_CREATE_TABLE + NEW_LINE + STATEMENT_ALTER_TABLE);
```
Comment (rock3r) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: `String.format()` would make this much more readable...

File: analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        );
+
+        String[] actual = getStatementsFromFile(file);
+
+        assertEquals(actual.length, 1);
+        assertEquals(actual[0], STATEMENT_FOR_COMMENTS_WITHIN_TESTS_1 + SPACE + STATEMENT_FOR_COMMENTS_WITHIN_TESTS_2);
+    }
+
+    private SQLFile givenSQLFileParsedFromString(String sql) throws IOException {
+        SQLFile file = new SQLFile();
```
Comment (rock3r) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: `SQLFile sqlFile`?

File: analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java (hunk 2 above)
Comment (frapontillo) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: It'd have to be hardcoded multiple times in 4 tests (2 in the "given" part, 2 in the "assert" part). The assert part is a bit more unstable, since it verifies that there is only one space in a specific position.

File: analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    private final static String NEW_LINE = "\n";
+    private final static String SPACE = " ";
+    private final static String[] EXPECTED_STATEMENTS = new String[]{STATEMENT_CREATE_TABLE, STATEMENT_ALTER_TABLE};
+    private final static String LINE_COMMENT = "-- this is a line comment";
+    private static final String BLOCK_COMMENT = "/* this is a \n block comment \n spanning over \n multiple lines */";
+    private final static String STATEMENT_FOR_COMMENTS_WITHIN_TESTS_1 = "CREATE TABLE";
+    private final static String STATEMENT_FOR_COMMENTS_WITHIN_TESTS_2 = "banana(_id INTEGER);";
+
+    @Test
+    public void givenSQLFile_whenParseAndGetStatements_thenGetCorrectStatements() throws IOException {
```
Comment (rock3r) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: Are you checking the `parse` or the `getStatements`? It looks like all your tests are testing two things, with this naming.

Comment (frapontillo) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: `parse` simply builds the statements, a private field that can only be accessed by `getStatements`.

File: analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java (hunk 4 above)
Comment (frapontillo) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: Not currently available in the repo.

File: analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java (hunk 5 above)
Comment (frapontillo) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: You mean like `givenSQLFileParsedFromString(String.format("%s\n%s", STATEMENT_CREATE_TABLE, STATEMENT_ALTER_TABLE));` to get rid of `NEW_LINE`, `SPACE`, etc?

File: analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java (hunk 6 above)
Comment (frapontillo) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: 319371007df69f501288b0086614dce588bed07a

File: analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java (hunk 7 above)
Comment (rock3r) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: Then you're testing the parsing, not the getter, right?

File: analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java (hunk 5 above)
Comment (rock3r) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: Yup

File: analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java (hunk 4 above)
Comment (rock3r) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: Oh well.

File: analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java (hunk 7 above)
Comment (frapontillo) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: Yup, fixed in 89b654577e5caf7623fea06ef692505472833adf

File: analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java (hunk 5 above)
Comment (frapontillo) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: I wouldn't agree that it makes it more readable. The only point in using constants is to avoid repetition, otherwise the best thing (in test) would be to explicitly define every input arguments:

```java
givenSQLFileParsedFromString(
    "CREATE TABLE banana(_id INTEGER);\n" +
    "ALTER TABLE banana ADD COLUMN colour TEXT;"
):
```

Both ways make reading clearer **and** allows sequential parsing (line by line) of the interested items, which is exactly what the `SQLFile` class does.

File: analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java (hunk 6 above)
Comment (takecare) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: shouldn't it change for the other methods too then?

**Comment by rock3r** [comment]
LGTM 👍 too

File: analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java (hunk 6 above)
Comment (frapontillo) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: ops, I did it again

aac255268a0ec3c3d10e9e4bcc08399c684878a4

Comment (takecare) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: aah young britney... what would you say if you met current britney?

File: analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java (hunk 2 above)
Comment (takecare) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: fair enough, but what do you mean by verifying the space is in a specific position? also, couldn't we extract the assert part then?

Comment (frapontillo) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: Each statement spanning over multiple line will be built on one line, concatenating with a single space. Asserting by manually concatenating space instances (`" "`) could trick the occasional developer into thinking that the space is just "any character".

With regards to extraction, what could be has already been extracted, see `EXPECTED_STATEMENTS`.
`STATEMENT_FOR_COMMENTS_WITHIN_TESTS_1 + SPACE + STATEMENT_FOR_COMMENTS_WITHIN_TESTS_2` remains twice in order to make it explicit that the statement, previously separated by a comment, is now concatenated with a space.

Comment (takecare) [inline analyzer/src/test/java/com/novoda/sqlite/impl/SQLFileTest.java]: 👍 thanks

**Comment by devisnik** [comment]
Nice!

Actually, `SQLFile` has its roots in the `sqliteProvider` project. It has deviated here somewhat already. Should we try to align across the two projects again?

**Comment by frapontillo** [comment]
Hmmm yeah we could do it, please take a look at #49 as well
